### ISSUE TITLE
[RMM-11] Fix REINFORCE gradient double-application and missing batch averaging

### ISSUE THREAD

**Issue by SkrOYC** [SkrOYC/langchain-middlewares-callbacks-ts#73]
**Parent:** #71 [RMM-9]
**Priority:** P1 — Core learning mechanism produces incorrect gradients
**Scope:** `src/middleware/hooks/after-model.ts`

## Problem

The REINFORCE gradient computation at `after-model.ts` has two mathematical errors that cause the reranker to learn incorrectly.

### Error 1: Double softmax gradient

The code applies BOTH equivalent forms of the softmax derivative simultaneously:

**Current code (conceptual, per memory i at position [r][c]):**

```
gradWq[r][c] += η * (1/τ) * (R_i - b) * (δ_i - P_i) * (m'_i[r] - E[m'[r]]) * q[c]
                                           ^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^
                                           Form A          Form B (equivalent to A)
```

**Mathematical derivation:**

For a softmax distribution, the gradient of `log P_i` with respect to score `s_j` is:

```
∂ log P_i / ∂ s_j = δ_{ij} - P_j         ... (Form A: score function)
```

Chain-ruled through `W_q` (since `∂s_j/∂W_q[r][c] = m'_j[r] * q[c]`):

```
∂ log P_i / ∂ W_q[r][c] = (1/τ) * Σ_j (δ_{ij} - P_j) * m'_j[r] * q[c]
                         = (1/τ) * (m'_i[r] - E_P[m'[r]]) * q[c]    ... (Form B: baseline)
```

Form A and Form B are the **same expression**. The code multiplies them together, creating a "squared" gradient that does not correspond to any valid policy gradient.

**For W_m, the same pattern:**

```
∂ log P_i / ∂ W_m[a][b] = (1/τ) * Σ_j (δ_{ij} - P_j) * q'[a] * m_j[b]
                         = (1/τ) * q'[a] * (m_i[b] - E_P[m[b]])
```

Note: `q'` (adapted query) and original `m_i` are the correct factors for W_m — this part the code gets right. The error is only the doubled application.

### Error 2: Missing batch averaging

**Current:** Gradients from each sample are summed into the accumulator.
**Expected:** Gradients should be averaged (divided by `batchSize`) per Equation 3.

```typescript
// Current (after-model.ts, gradient accumulation loop)
accumulator.accumulatedGradWq = addMatrix(
  accumulator.accumulatedGradWq,
  sampleGradients.gradWq  // ← Summed, not averaged
);
```

**Impact:** Effective learning rate scales with batch size. With `batchSize=4`, the effective η is 4x the configured value.

### Fix 1: Choose one gradient form

**Recommended: Keep Form B (baseline subtraction), remove `coef`.**

The baseline form `(m'_i[r] - E[m'[r]])` is cleaner because it naturally computes `∂ log P_i / ∂ W_q` without needing to iterate over all j. The per-memory loop already computes `expectedMemAdapted` and `expectedMemOriginal`.

For W_q at `[r][c]`, the correct gradient contribution from a turn with reward R is:

```
grad_Wq[r][c] = η * (1/τ) * (R - b) * (m'_i[r] - E[m'[r]]) * q[c]
```

For W_m at `[a][b]`:

```
grad_Wm[a][b] = η * (1/τ) * (R - b) * q'[a] * (m_i[b] - E[m[b]])
```

Where the sum is over selected memories i ∈ M_M.

### Fix 2: Add batch averaging

```typescript
const averagedGradWq = scaleMatrix(sampleGradients.gradWq, 1 / batchSize);
accumulator.accumulatedGradWq = addMatrix(
  accumulator.accumulatedGradWq,
  averagedGradWq
);
```

### Testing strategy

Create unit tests with small (2x2 or 3x3) matrices where gradients can be hand-computed:

1. Single memory, single selection → verify gradient matches analytic formula
2. Two memories, one selected → verify `(δ_i - P_i)` OR `(m'_i - E[m'])` form, not both
3. Batch of 2 samples → verify averaging divides by 2
4. Zero advantage (R = b) → verify zero gradient

## Acceptance Criteria

- [ ] Gradient formula uses ONE softmax derivative form, not both
- [ ] Batch averaging divides accumulated gradients by `batchSize`
- [ ] Unit tests verify gradient values against hand-computed examples (at least 4 test cases)
- [ ] All 559 existing tests pass
- [ ] W_q gradient uses adapted memory `m'_i` and original query `q` (unchanged)
- [ ] W_m gradient uses adapted query `q'` and original memory `m_i` (unchanged)

## References

- Paper Equation 3: `Δφ = η · (R - b) · ∇_φ log P(M_M | q, M_K; φ)`
- Paper Appendix A.1: η = 1e-3, b = 0.5, batch_size = 4
- SPEC.md Section 3.3.3
